### PR TITLE
Facilitate customization of setupTest* functions

### PR DESCRIPTION
**Pull request by kategengler**
[Rendered](https://github.com/kategengler/rfcs-1/blob/test-setup/text/0637-customizable-test-setups.md)

### CODE REVIEW

File: text/0637-customizable-test-setups.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+# Facilitate customization of setupTest* functions
+
+## Summary
+
+Provide a default and convenient way for each project to customize the
+`setupApplicationTest`, `setupRenderingTest`, and `setupTest` functions from
+[RFC #268](https://github.com/emberjs/rfcs/blob/master/text/0268-acceptance-testing-refactor.md)
+and [RFC #232](https://github.com/emberjs/rfcs/blob/master/text/0232-simplify-qunit-testing-api.md).
+The app and addon blueprints will be updated to create a file at `
+tests/helpers/index.js` where those functions will be wrapped and exported,
```
Comment (MelSumner) [inline text/0637-customizable-test-setups.md:15]:
```suggestion
The app and addon blueprints will be updated to create a file at 
`tests/helpers/index.js` where those functions will be wrapped and exported, 
```

Comment (kategengler) [reply text/0637-customizable-test-setups.md]: Thanks!

File: text/0637-customizable-test-setups.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+changes and waste time to debug test failures or weird test side effects.
+
+## Detailed design
+
+When generating tests in an Ember project today, the setup functions in the
+generated test are imported from `ember-qunit` or `ember-mocha`. This RFC
+proposes adding a file at `test/helpers/index.js` that would look like:
+
+```js
+import { setupApplication as upstreamSetupApplicationTest } from 'ember-qunit';
```
Comment (pgengler) [inline text/0637-customizable-test-setups.md:91]:
```suggestion
import { setupApplicationTest as upstreamSetupApplicationTest } from 'ember-qunit';
```

**Comment by ro0gr** [comment]
What would be an import part within addons?

Is it an absolute path:
```js
    import setupTest from 'dummy/tests/helpers';
```
Or relative:
```js
    import setupTest from '../tests/helpers';
```

**Comment by jgwhite** [comment]
This is terrific. It seems plausible that addons could use this new convention to add their `setup*` functions automatically when installed. For example, `ember install ember-cli-mirage` could codemod `tests/helpers/index.js` as follows:

```diff
  import { setupApplicationTest as upstreamSetupApplicationTest } from 'ember-qunit';
+ import { setupMirage } from 'ember-cli-mirage/test-support';
  export function setupApplicationTest(hooks, options) {
    upstreamSetupApplicationTest(hooks, options);
+   setupMirage(hooks);
    // customize here
  }
  export { setupTest, setupRenderingTest } from 'ember-qunit';
```

**Comment by rwjblue** [comment]
@jgwhite 

> This is terrific. It seems plausible that addons could use this new convention to add their setup* functions automatically when installed. For example, ember install ember-cli-mirage could codemod tests/helpers/index.js as follows...

Yes, exactly.

**Comment by rwjblue** [comment]
@ro0gr 

> What would be an import path within addons?

It would be absolute.

**Comment by kategengler** [comment]
We discussed this in today's framework meeting and are moving it into Final Comment Period.

File: text/0637-customizable-test-setups.md [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+
+## Motivation
+
+Projects often need to customize setup for every test. For example, in 
+acceptance tests, it is often necessary to override services, mock APIs, etc.
+in every test of that type.
+
+A common use-case is `ember-cli-mirage`, which provides a `setupMirage` function
+to be called after `setupAcceptanceTest`. In a project using this, it can be
```
Comment (Turbo87) [inline text/0637-customizable-test-setups.md:27]:
```suggestion
to be called after `setupApplicationTest`. In a project using this, it can be 
```

Comment (kategengler) [reply text/0637-customizable-test-setups.md]: Good catch, thanks!

**Comment by sandstrom** [comment]
This would be very useful! 💯

**Comment by rwjblue** [comment]
🎉

**Comment by lougreenwood** [comment]
Are there any plans for implementing this?

We adopted this format for our project about 6 months ago, but the burden of updating the imports for generated tests is beginning to get annoying. We're considering whether to just update our local blueprints or possibly chip in on the RFC implementation if it isn't too complex.

**Comment by rwjblue** [comment]
> Are there any plans for implementing this?

As far as I know, this hasn't been started by anyone. I don't think the specific implementation is terribly difficult (though perhaps there is something we missed in the initial design), and I'd be happy to help guide you through the steps if you want to pick it up (just ping me in discord).

**Comment by bertdeblock** [comment]
@lougreenwood I have been working on this. The corresponding PRs are listed above.

**Pull request merged**